Convert FunctionsOfTime to cache tag

osheamonn opened the pull request:
### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

markscheel committed:
ActionTesting: Allow one to specify MutableGlobalCache items.

kidder commented on src/Evolution/Initialization/DgDomain.hpp:
```diff
@@ -53,14 +53,13 @@ namespace Initialization {
  * - Uses:
  *   - `domain::Tags::InitialExtents<Dim>`
- *   - `domain::Tags::InitialFunctionsOfTime<Dim>`
+ *   - `domain::Tags::FunctionsOfTime<Dim>`
```
the tag is no longer templated on the dimension

kidder commented on tests/Unit/Framework/ActionTesting.hpp:
```diff
@@ -1662,13 +1664,39 @@ class MockRuntimeSystem {
 
+  /// Construct from the tuple of GlobalCache objects and MutableGlobalCache
+  /// objects.
+  explicit MockRuntimeSystem(CacheTuple cache_contents,
```
should not be marked explicit (as it takes two arguments...)

kidder replied:
and can we just have the previous constructor call this one?

kidder commented on tests/Unit/Framework/ActionTesting.hpp:
```diff
@@ -1662,13 +1664,39 @@ class MockRuntimeSystem {
+  /// Construct from the tuple of GlobalCache and MutableGlobalCache
+  /// objects that might be in a different order.
+  template <typename... CacheTags, typename... MutableCacheTags>
+  explicit MockRuntimeSystem(
```
remove the explicit

kidder commented on tests/Unit/Domain/Test_TagsTimeDependent.cpp:
```diff
@@ -38,8 +38,6 @@
 void test_tags() noexcept {
-  TestHelpers::db::test_simple_tag<domain::Tags::InitialFunctionsOfTime<Dim>>(
-      "InitialFunctionsOfTime");
   TestHelpers::db::test_simple_tag<domain::Tags::FunctionsOfTime>(
```
please remove this test as well; it is duplicated in tests/Unit/Domain/FunctionsOfTime/Test_Tags.cpp

kidder commented on src/Parallel/Algorithm.hpp:
```diff
@@ -354,7 +354,11 @@ class AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>> {
   bool terminate_{true};
 
-  using all_cache_tags = get_const_global_cache_tags<metavariables>;
+  using const_cache_tags = get_const_global_cache_tags<metavariables>;
```
We actually don't want to be able to get mutable_cache_tags from the DataBox as you need to check if they are valid.  So this change should be dropped unless I am missing some use of the new aliases, in which case only the `const_cache_tags` should be passed into `db::wrap_tags_in<Tags::FromGlobalCache, all_cache_tags>` in the two places it occurs

osheamonn replied:
I did this because the compute tag CoordinatesMeshVelocityAndJacobiansCompute takes FunctionsOfTime as an argument tag. Disallowing mutable cache tags to be accessible from the databox means that tag can no longer be used as a compute tag, and updating its value will have to be explicitly done as its own action or a part of another action. I deliberately didn't want to do this since it changed more existing code, but if you think that's how this tag should be handled then that change will have to be made.

kidder replied:
I discussed this with @nilsdeppe and we still don't want to be able to get mutable cache tags from the DataBox as there is no mechanism for checking if they are in a valid state.  Therefore, like you say, any compute tag that takes FunctionsOfTime will need to be replaced with a simple tag that needs to be mutated in a separate action or part of another action.  For now the simplest way is probably to do so in its own action that would be called right after the time is updated.  If you want to discuss this in further detail, please let us know and we would be happy to hop into a hangout.

kidder replied:
This was discussed again last week at a meeting that included @nilsdeppe @wthrowe @osheamonn @markscheel @moxcodes @fmahebert.  There we decided that we do want `FunctionsOfTime` accessible through the `DataBox`, but instead of adding all mutable cache tags, for the time being we will only add the `FunctionsOfTime`.

kidder commented on tests/Unit/Framework/ActionTesting.hpp:
```diff
@@ -673,7 +673,10 @@ class MockDistributedObject {
+  using const_cache_tags = Parallel::get_const_global_cache_tags<metavariables>;
+  using mutable_cache_tags =
+      Parallel::get_mutable_global_cache_tags<metavariables>;
+  using all_cache_tags = tmpl::append<const_cache_tags, mutable_cache_tags>;
```
I believe the same comment I made above about not placing the mutable_cache_tags in the DataBox apply here as well

osheamonn committed:
Convert FunctionsOfTime to mutable cache tag

osheamonn commented:
Updated.

I added the FunctionsOfTime to the DataBox using the `FromGlobalCache`compute tag. This has a bit of a knock-on effect for the testing of the time dependent tags, since `MeshVelocity..` etc. depends on `FromGlobalCache<FunctionsOfTime>`, which itself would depend on global cache. To avoid having to change these tests to require constructing a global cache, I added a tag to mock `FromGlobalCache` in tests/Helpers/Parallel/MockFromGlobalCache.hpp. 

I couldn't think of a better way to achieve this which wouldn't require a lot of code changes.

kidder commented on src/Domain/TagsTimeDependent.hpp:
```diff
@@ -77,7 +86,8 @@ struct CoordinatesMeshVelocityAndJacobiansCompute
       tmpl::list<MapTagGridToInertial, Tags::Coordinates<dim, Frame::Grid>,
-                 ::Tags::Time, Tags::FunctionsOfTime>;
+                 ::Tags::Time,
+                 Parallel::Tags::FromGlobalCache<Tags::FunctionsOfTime>>;
```
you should still be able to just specify `Tags::FunctionsOfTime` since the `Parallel::Tags::FromGlobalCache<Tags::FunctionsOfTime>` is derived from it

kidder replied:
and then this should make it unnecessary to have your mock FromGlobalCache tag in the tests

osheamonn replied:
Great thanks! I temporarily forgot how base tags work.

kidder commented on tests/Unit/Domain/Test_TagsTimeDependent.cpp:
```diff
@@ -122,7 +119,7 @@ void test() noexcept {
       Tags::Time, domain::Tags::Coordinates<Dim, Frame::Grid>,
       domain::Tags::InverseJacobian<Dim, Frame::Logical, Frame::Grid>,
-      domain::Tags::FunctionsOfTime,
+      Parallel::Tags::FromGlobalCache<domain::Tags::FunctionsOfTime>,
```
see above comment, this change should not be needed

kidder commented on tests/Unit/Evolution/Test_TagsDomain.cpp:
```diff
@@ -102,7 +103,7 @@ void test() noexcept {
       domain::Tags::Coordinates<Dim, Frame::Grid>,
       domain::Tags::InverseJacobian<Dim, Frame::Logical, Frame::Grid>,
-      domain::Tags::FunctionsOfTime,
+      Parallel::Tags::FromGlobalCache<domain::Tags::FunctionsOfTime>,
```
again, this change should not be needed

kidder commented on src/Domain/TagsTimeDependent.hpp:
```diff
@@ -23,6 +23,7 @@
 #include "Utilities/Gsl.hpp"
 #include "Utilities/TMPL.hpp"
 
+
```
remove this

kidder reviewed (changes requested):
just one blank line to delete (you can squash)
